File: code/calc_mahalanobis_fn.R
# The MD/sigma dissimilarity function
# Testing: pt_i  <- 1000; chunk  <- chunks[[100]]
# Core Mahalanobis distance -> sigma calculation
calc_mahalanobis <- function(
    pt_i,
    focal_data_cov,
    focal_data_mean,
    analog_data,
    var_names,
    n_analog_pool,
    n_analog_use,
    min_dist
) {
    # Build cov matrix for pt_i from 30 years of annual projected future data
    cov_i <- purrr::map(focal_data_cov, ~ .x[pt_i]) |>
        rbindlist() |>
        dplyr::select(-x, -y) |>
        cov()

    # Calculate the mean of the future annuals -----------------------------------
    # Option to supply these mean focal data rather than derive (in the case of contemporary validation)
    if (length(focal_data_mean) == 1 ) {
        focal_data_mean <- focal_data_mean[pt_i] |>
            select(all_of(var_names)) |>
            unlist()
    } else {
        focal_data_mean <- purrr::map(focal_data_cov, ~ .x[pt_i]) |>
            rbindlist() |>
            # Calculate mean of 30 years for every point (row in data.table)
            dplyr::summarise(across(all_of(var_names), mean)) |>
            unlist()
    }

    n_analog_pool = 1000000
    tic()
    # Analog pool (from historical normals) ------------------------------------
    # Random global sample == n_analog_pool
    random_pts <- seq_len(nrow(analog_data)) |>
        sample(size = n_analog_pool)

    # Build reference matrix from random sample of historical normals
    analog_mat <- analog_data[random_pts, ] |>
        dplyr::select(-x, -y) |>
        as.matrix()

    # Sigma dissimilarity between pt_i and analog pool ---------------------------
    # Calculate Mahalanobis distances
    d <- mahalanobis(analog_mat, focal_data_mean, cov_i)
    toc()

    # Save output
    out_dt <- data.table(
        "f_x" = focal_data_cov[[1]][pt_i][["x"]], 
        "f_y" = focal_data_cov[[1]][pt_i][["y"]],
        "a_x" = analog_data[random_pts][["x"]],
        "a_y" = analog_data[random_pts][["y"]],
        "md" = d
    )

    out_dt <- out_dt |>
        # Euclidean/geographic distance between focal point and analogs
        dplyr::mutate("dist_km" = round(sqrt((f_x - a_x)^2 + (f_y - a_y)^2)*0.001, 1)) |>
        dplyr::filter(dist_km > min_dist) |>
        # Sort all analogs by sigma
        dplyr::arrange(md) |>
        # Pick n analogs from pool with lowest sigma value
        dplyr::slice_head(n = n_analog_use) |>
        # Other speed improvement: only calculate sigma for the analogs we're keeping!
        # Can order the best 1000 based on D... relative order is the same as sigma.
        dplyr::mutate(sigma = calc_sigma(md, length(focal_data_mean))) |>
        dplyr::mutate(sigma = round(sigma, 4),
                      md = round(md, 3))

    rm(cov_i, focal_data_mean, random_pts, analog_mat, d)
    # This takes ~50% of compute time... not sure if neccesary to prevent memory leakge
    gc()
    return(out_dt)
}

File: code/src/climate_analogs_jl/calc_mahalanobis_fn.jl
using DataFrames
using DataFramesMeta
using Distributions
using Distances
using Suppressor
using Random
using CSV
using Statistics
using StatsBase
using LinearAlgebra
using ThreadPools: @bthreads

#fast sort from <https://discourse.julialang.org/t/faster-sorting-of-dataframes-jl-via-background-ordering/45084>
assignnew!(v, idx) = v .= v[idx]
function fsort!(dataframe, by)
	idx = sortperm(dataframe[!, by])

	@bthreads for n in names(dataframe)
		assignnew!(dataframe[!, n], idx)
	end
	dataframe
end


#function for inverse covariance matrix
function calculate_covariance_matrix(pt_i::Integer, focal_data_cov::Union{Vector{Any}, Vector{DataFrame}})
	mat::Matrix{Int16} =
		DataFrame([focal_data_cov[i][pt_i, :] for i in 1:length(focal_data_cov)]) |>
		#select only the variables we care about
		df -> select(df, Not(:x, :y)) |>
			  Matrix
	#check if a column is all the same value, find 
	check_cols = [all(mat[:, i] .== mat[1, i]) for i in 1:size(mat, 2)]


	# compute the covariance matrix
	cov_i::Matrix{Float32} = cov(mat)

	# if any columns are all the same value, add Tikhonov regularization
	if any(check_cols)
		cov_i = cov_i + 0.0001 * I
	end

	inv_i::Matrix{Float32} = inv(cov_i)

	return inv_i
end
#function for Mahalanobis distance
function calculate_sqmahalanobis_distance(
	cov_i::AbstractMatrix{T1},
	analog_mat::AbstractMatrix{T2},
	focal_data_mean_i::AbstractMatrix{T3}) where {T1 <: Real, T2 <: Real, T3 <: Real}

	d::Vector{Float32} = pairwise(SqMahalanobis(cov_i, skipchecks = true), analog_mat, focal_data_mean_i, dims = 1) |>
						 vec

	return d
end

function process_mahalanobis_data(
	insert_df::DataFrame,
	f_x::AbstractFloat,
	f_y::AbstractFloat,
	min_dist::Union{AbstractFloat, Integer},
	max_dist::Union{AbstractFloat, Integer},
	n_analog_use::Integer,
	focal_data_mean_i::Matrix{Float32},
)

	# Calculate great circle distance and round to 1 decimal place

	insert_df.dist_km = round.(great_circle_distance(insert_df.a_y, insert_df.a_x, f_y, f_x), digits = 1) #24.1ms

	# Subset the DataFrame based on the minimum distance
	subset!(insert_df, :dist_km => ByRow(>(min_dist))) # 5 ms

	# Subset the DataFrame based on maximum distance
	subset!(insert_df, :dist_km => ByRow(<(max_dist))) # 5 ms

	# Sort the DataFrame by Mahalanobis distance
	fsort!(insert_df, :md) #5 ms 17 allocations

	# Select the top n_analog_use rows
	@inbounds result_i_sub = insert_df[1:n_analog_use, :] # 1.9 μs 26 allocations
	insert_df = nothing
	# Calculate sigma and round to 4 decimal places
	result_i_sub.sigma = calc_sigma.(result_i_sub.md, length(focal_data_mean_i)) |> df -> round.(df, digits = 4) #38 μ s 104 allocations

	# Round Mahalanobis distance to 3 decimal places
	result_i_sub.md = round.(result_i_sub.md, digits = 3)

	# Repeat f_x and f_y for each row in the DataFrame
	result_i_sub.f_x = repeat([f_x], nrow(result_i_sub))
	result_i_sub.f_y = repeat([f_y], nrow(result_i_sub))

	return result_i_sub
end


function calc_mahalanobis(
	pt_i::Integer,
	focal_data_cov::Union{Vector{Any}, Vector{DataFrame}},
	focal_data_mean::DataFrame,
	var_names::Vector{String},
	n_analog_use::Integer,
	min_dist::Union{AbstractFloat, Integer},
	max_dist::Union{AbstractFloat, Integer},
	input_sample::DataFrame,
	insert_df::DataFrame,
)
	# Build cov matrix for pt_i from 30 years of annual projected future data


	cov_i = calculate_covariance_matrix(pt_i, focal_data_cov) #50 μs 306 allocations




	# Calculate the mean of the future annuals -----------------------------------
	# Option to supply these mean focal data rather than derive (in the case of contemporary validation)
	#44μs 288 allocations
	if size(focal_data_mean, 2) == 1
		focal_data_mean_i = focal_data_mean[pt_i][var_names]
	else
		focal_data_mean_i::Matrix{Float32} = DataFrame([focal_data_cov[i][pt_i, :] for i in 1:length(focal_data_cov)]) |>
											 df -> select(df, Not(:x, :y)) |>
												   df -> mean(Matrix(df), dims = 1)
	end



	analog_mat = Matrix(input_sample[:, var_names]) #910 μs 10 allocations / 1.9 ms 37 allocations

	# Calculate the sigma dissimilarity -----------------------------------------


	# Sigma dissimilarity between pt_i and analog pool ---------------------------
	# Calculate Mahalanobis distances from analog mat to focal_data_mean



	insert_df.md = @suppress calculate_sqmahalanobis_distance(cov_i, analog_mat, focal_data_mean_i) #32 ms 249 allocations
	cov_i = nothing
	analog_mat = nothing

	f_x = focal_data_cov[1][pt_i, "x"] # 100 ns 1 allocation
	f_y = focal_data_cov[1][pt_i, "y"] # 123 ns 1 allocation
	# Save output

	# script to here is 35.188 ms 881 allocations

	result_i_final = process_mahalanobis_data(insert_df, f_x, f_y, min_dist, max_dist, n_analog_use, focal_data_mean_i) # 32 ms 394 allocations

	f_x = nothing
	f_y = nothing
	min_dist = nothing
	n_analog_use = nothing

	return result_i_final
end

